File: backend/app/api/custom.py
"""Custom rule management API."""
from typing import List
import logging

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.services.custom_recognizer_service import CustomRecognizerService


logger = logging.getLogger(__name__)

router = APIRouter(prefix="/custom", tags=["custom"])
custom_service = CustomRecognizerService()


class AddKeywordRequest(BaseModel):
    entity_type: str
    keywords: List[str]
    score: float = 0.95
    description: str = ""


class AddPatternRequest(BaseModel):
    entity_type: str
    regex: str
    context: List[str] = Field(default_factory=list)
    score: float = 0.9
    description: str = ""


@router.get("/config")
async def get_config():
    """Return the current custom keyword and pattern configuration."""
    try:
        custom_service.load_config()
        return {
            "keywords": custom_service.custom_keywords,
            "patterns": custom_service.custom_patterns,
            "keywords_count": len(custom_service.custom_keywords),
            "patterns_count": len(custom_service.custom_patterns),
        }
    except Exception as exc:
        logger.error("Failed to load custom configuration: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.post("/keywords")
async def add_keywords(request: AddKeywordRequest):
    """Create or replace a keyword-based custom rule."""
    try:
        custom_service.add_keyword_rule(
            entity_type=request.entity_type,
            keywords=request.keywords,
            score=request.score,
            description=request.description,
        )
        custom_service.save_config()
        return {
            "status": "success",
            "message": f"已保存 {request.entity_type} 关键词规则",
        }
    except Exception as exc:
        logger.error("Failed to save keyword rule: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.post("/patterns")
async def add_pattern(request: AddPatternRequest):
    """Create or replace a regex-based custom rule."""
    try:
        custom_service.add_pattern_rule(
            entity_type=request.entity_type,
            regex=request.regex,
            context=request.context,
            score=request.score,
            description=request.description,
        )
        custom_service.save_config()
        return {
            "status": "success",
            "message": f"已保存 {request.entity_type} 正则规则",
        }
    except Exception as exc:
        logger.error("Failed to save pattern rule: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.delete("/keywords/{entity_type}")
async def delete_keywords(entity_type: str):
    """Delete all keywords for the given entity type."""
    try:
        if not custom_service.delete_keyword_rule(entity_type):
            raise HTTPException(status_code=404, detail="关键词规则不存在")

        custom_service.save_config()
        return {
            "status": "success",
            "message": f"已删除 {entity_type} 关键词规则",
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("Failed to delete keyword rule: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.delete("/patterns/{entity_type}")
async def delete_pattern(entity_type: str):
    """Delete the regex rule for the given entity type."""
    try:
        if not custom_service.delete_pattern_rule(entity_type):
            raise HTTPException(status_code=404, detail="正则规则不存在")

        custom_service.save_config()
        return {
            "status": "success",
            "message": f"已删除 {entity_type} 正则规则",
        }
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("Failed to delete pattern rule: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.post("/reload")
async def reload_config():
    """Reload custom rules from disk."""
    try:
        custom_service.load_config()
        return {
            "status": "success",
            "message": "配置已重新加载",
            "keywords_count": len(custom_service.custom_keywords),
            "patterns_count": len(custom_service.custom_patterns),
        }
    except Exception as exc:
        logger.error("Failed to reload custom configuration: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))


@router.get("/test")
async def test_custom_recognizer(text: str):
    """Run custom rule recognition against the supplied text."""
    try:
        custom_service.load_config()
        entities = custom_service.match_all(text)
        return {
            "text": text,
            "entities": entities,
            "count": len(entities),
        }
    except Exception as exc:
        logger.error("Failed to test custom recognizer: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=str(exc))

File: frontend/src/api/custom.ts
import request from '@/utils/request'

export interface KeywordRule {
  description?: string
  keywords: string[]
  score: number
}

export interface PatternRule {
  name: string
  description?: string
  regex: string
  score: number
  context?: string[]
}

export interface CustomConfig {
  keywords: Record<string, KeywordRule>
  patterns: PatternRule[]
  keywords_count: number
  patterns_count: number
}

export interface AddKeywordRequest {
  entity_type: string
  keywords: string[]
  score: number
  description?: string
}

export interface AddPatternRequest {
  entity_type: string
  regex: string
  context: string[]
  score: number
  description?: string
}

export interface TestRecognizerResponse {
  text: string
  count: number
  entities: Array<{
    type: string
    text: string
    start: number
    end: number
    score: number
    source: string
    metadata?: Record<string, unknown>
  }>
}

export const getCustomConfig = () => {
  return request.get<any, CustomConfig>('/custom/config')
}

export const addKeywords = (data: AddKeywordRequest) => {
  return request.post('/custom/keywords', data)
}

export const addPattern = (data: AddPatternRequest) => {
  return request.post('/custom/patterns', data)
}

export const deleteKeywords = (entityType: string) => {
  return request.delete(`/custom/keywords/${entityType}`)
}

export const deletePattern = (entityType: string) => {
  return request.delete(`/custom/patterns/${entityType}`)
}

export const reloadConfig = () => {
  return request.post('/custom/reload')
}

export const testRecognizer = (text: string) => {
  return request.get<any, TestRecognizerResponse>('/custom/test', {
    params: { text }
  })
}
